### PR TITLE
Switch to bank permission system

### LINKED ISSUE #31: Switch to 'Bank Permissions' system.
Suggested by @MrNemo64.

See [this](https://www.spigotmc.org/threads/new-economy-api-looking-for-your-feedback.527805/page-3#post-4280097) reply on SpigotMC.

### PR DESCRIPTION
**Pull request by MrNemo64**
PR to close #31

### CODE REVIEW

**Comment by MrNemo64** [comment]
You may start leaving some feedback

I havent made `canAddMembers`, `canAddOwners`... methods because of what I said [here](https://github.com/lokka30/Treasury/issues/31#issuecomment-988306000)

> At this point we need to define what a member and what an owner is on a bank account. Is a member just someone with WITHDRAW DEPOSIT and CONSULT permissions? Is an owner just a player with all permissions?

File: api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java [hunk 1]
```diff
@@ -24,89 +24,137 @@
  * balance shared by multiple Bank Members, such as a 'town balance'
  * or 'faction balance'. The linked Player can be changed at any
  * time (Owning Player).
- *
  * @author lokka30, MrNemo64
  * @see Account
  * @since v1.0.0
  */
-@SuppressWarnings({"unused"})
+@SuppressWarnings({ "unused" })
```
Comment (lokka30) [inline api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java:31]: I'm unsure whether we should remove these warning suppressions from the project, and instead disable it for Treasury specifically in our IDE (disable inspection button).

File: api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java [hunk 2]
```diff
@@ -24,89 +24,137 @@
...
+     * Checks whether the given player can consult the balance of this bank account,
+     * in other words, if they have the {@link BankAccountPermission#CONSULT} permission.
+     * @param player the {@link UUID} of the player to check if they can consult the balance
+     * @param subscription the {@link EconomySubscriber} accepting whether the player can consult the balance
+     * @author MrNemo64
+     * @see #hasPermission(UUID, BankAccountPermission, EconomySubscriber)
+     * @see BankAccountPermission#CONSULT
+     * @since v1.0.0
+     */
+    default void canConsultBalance(@NotNull UUID player, @NotNull EconomySubscriber<Boolean> subscription) {
```
Comment (lokka30) [inline api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java:128]: I don't believe we should have the methods for individual permissions - rather, just the one that asks 'has permission' with specified enum constant.

File: api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccountPermission.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ */
+public enum BankAccountPermission {
+
+    /**
+     * Allows a player to consult the balance of the {@link BankAccount}.
+     * @see Account#retrieveBalance(Currency, EconomySubscriber)
+     * @since v1.0.0
+     */
+    CONSULT,
+    /**
```
Comment (lokka30) [inline api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccountPermission.java:34]: I think we should have an empty line between each constant and javadoc comment

**Review by lokka30 — CHANGES_REQUESTED** [review]
Thank you very much for the PR @MrNemo64, here's what I could spot for change. The rest looks very good. Please let me know what you think of my suggestions :)

File: api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java (hunk 1 above)
Comment (MrIvanPlays) [reply api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java:31]: Remove all `@SuppressWarnings` is the go for a cleaner code.

Comment (lokka30) [reply api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java:31]: My suggestion here is outside of the scope of this PR, I've instead created an issue which I will work on once this PR is merged.

File: api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccountPermission.java (hunk 3 above)
Comment (MrNemo64) [reply api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccountPermission.java:34]: So you mean
```Java
/**
 * ...
 */
ENUM_ENTRY,

/**
 * ...
 */
ENUM_ENTRY
```
?

Comment (MrIvanPlays) [reply api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccountPermission.java:34]: yes exactly what he means

**Review by MrIvanPlays — APPROVED** [review]

**Review by lokka30 — APPROVED** [review]

File: api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java (hunk 2 above)
Comment (lokka30) [reply api/src/main/java/me/lokka30/treasury/api/economy/account/BankAccount.java]: @MrNemo64 you marked this as resolved but didn't leave a comment - what do you think about this?

**Comment by lokka30** [comment]
Thank you very much

**Pull request merged**